File: component/balancer/balancer.go
// Package balancer implements the `balancer` component: a controller-
// driven crossfade between its `left` and `right` inputs. Mirrors
// ondes.synth.wire.Balancer.
//
//	lScale = (ctrl/ctrlInputAmp + 1) / 2     // 0..1 as ctrl sweeps -amp..+amp
//	out    = (lScale*left + (1-lScale)*right) * level-scale
//
// With no `ctrl` input connected the sum is 0, giving lScale 0.5 - an
// even blend of left and right.
//
// Scale note: ctrlInputAmp is configured in Java-int units; our ctrl
// wire is in float units (Java-int / 32767), so we divide by
// ctrlInputAmp/32767 to recover the int-domain ratio. The `initial-value`
// config is parsed for compatibility but, as in Java, never affects the
// output (ctrl comes entirely from the wired input).
package balancer

import (
	"ondes/component"
	"ondes/synth"
)

const intScale = 32767.0

func init() {
	component.Register("balancer", func() component.Component { return &Balancer{} })
}

type Balancer struct {
	out         *synth.Wire
	namedInputs map[string][]*synth.Wire

	scale           float64
	ctrlInputAmpInv float64 // intScale / ctrlInputAmp
}

func (b *Balancer) Configure(spec component.Spec, v *synth.Voice, _ string) error {
	b.namedInputs = map[string][]*synth.Wire{}
	b.scale = 1
	ctrlInputAmp := 1000.0 // Java default

	if ls, ok := numeric(spec["level-scale"]); ok {
		b.scale = ls
	}
	// input-ctrl: { amp: <int>, initial-value: <float> }. Both keys must
	// be present for the override to take effect (matching getInAmpPair).
	if amp, _, ok := inAmpPair(spec, "input-ctrl", "initial-value"); ok {
		ctrlInputAmp = amp
	}
	b.ctrlInputAmpInv = intScale / ctrlInputAmp

	b.out = v.NewWire(b.compute)
	return nil
}

func (b *Balancer) Output() *synth.Wire { return b.out }

func (b *Balancer) AddInput(selectName string, w *synth.Wire) {
	if selectName == "" {
		selectName = "main"
	}
	b.namedInputs[selectName] = append(b.namedInputs[selectName], w)
}

func (b *Balancer) namedInputSum(pin string) float64 {
	var sum float64
	for _, w := range b.namedInputs[pin] {
		sum += w.Get()
	}
	return sum
}

func (b *Balancer) compute() float64 {
	l := b.namedInputSum("left")
	r := b.namedInputSum("right")
	ctrl := b.namedInputSum("ctrl")

	lScale := (ctrl*b.ctrlInputAmpInv + 1) / 2
	return (lScale*l + (1-lScale)*r) * b.scale
}

func numeric(v any) (float64, bool) {
	switch x := v.(type) {
	case int:
		return float64(x), true
	case int64:
		return float64(x), true
	case float64:
		return x, true
	}
	return 0, false
}

func inAmpPair(spec component.Spec, key, prop string) (amp, propVal float64, ok bool) {
	m, isMap := spec[key].(map[string]any)
	if !isMap {
		return 0, 0, false
	}
	amp, aok := numeric(m["amp"])
	propVal, pok := numeric(m[prop])
	if !aok || !pok {
		return 0, 0, false
	}
	return amp, propVal, true
}

File: c/src/component/balancer.c
// `balancer`: a controller-driven crossfade between `left` and `right`
// inputs. Ported from component/balancer/balancer.go.
//   lScale = (ctrl/ctrlInputAmp + 1) / 2
//   out    = (lScale*left + (1-lScale)*right) * level-scale
#include <string.h>

#include "computil.h"
#include "ondes/component.h"

static const double INT_SCALE = 32767.0;

typedef struct {
    Component base;
    Wire *out;
    WireList left, right, ctrl;
    double scale;
    double ctrl_input_amp_inv; // intScale / ctrlInputAmp
} Balancer;

static double balancer_compute(void *ctx) {
    Balancer *b = ctx;
    double l = wirelist_sum(&b->left);
    double r = wirelist_sum(&b->right);
    double c = wirelist_sum(&b->ctrl);
    double l_scale = (c * b->ctrl_input_amp_inv + 1) / 2;
    return (l_scale * l + (1 - l_scale) * r) * b->scale;
}

static Wire *balancer_output(Component *self) { return ((Balancer *)self)->out; }

static void balancer_add_input(Component *self, const char *sel, Wire *w) {
    Balancer *b = (Balancer *)self;
    if (sel && !strcmp(sel, "left"))
        wirelist_add(&b->left, w);
    else if (sel && !strcmp(sel, "right"))
        wirelist_add(&b->right, w);
    else if (sel && !strcmp(sel, "ctrl"))
        wirelist_add(&b->ctrl, w);
}

static int balancer_configure(Component *self, const Spec *spec, Voice *v,
                              const char *name) {
    (void)name;
    Balancer *b = (Balancer *)self;
    Arena *a = voice_arena(v);
    wirelist_init(&b->left, a);
    wirelist_init(&b->right, a);
    wirelist_init(&b->ctrl, a);
    b->scale = spec_double(spec, "level-scale", 1.0);

    double ctrl_input_amp = 1000.0; // Java default
    // input-ctrl: { amp, initial-value } - both keys required to override.
    const Spec *ic = spec_get(spec, "input-ctrl");
    double amp, iv;
    if (spec_is_map(ic) && spec_num(ic, "amp", &amp) &&
        spec_num(ic, "initial-value", &iv))
        ctrl_input_amp = amp;
    b->ctrl_input_amp_inv = INT_SCALE / ctrl_input_amp;

    b->out = voice_new_wire(v, balancer_compute, b);
    return 0;
}

static const ComponentVTable BALANCER_VT = {
    .configure = balancer_configure,
    .output = balancer_output,
    .add_input = balancer_add_input,
};

Component *balancer_new(Arena *a) {
    Balancer *b = arena_alloc(a, sizeof(*b));
    b->base.vt = &BALANCER_VT;
    return &b->base;
}
